enhancements/automated-service-ca-rotation.md

```diff
@@ -177,6 +183,22 @@ manual rotation (by deleting the signing secret) could be performed. Manual
 rotation would not be seamless, however, due to the lack of a grace period for
 services and clients to converge on the new trust chain.
 
+- Testing in advance of service CA rotation whether consumers of CA artifacts are
+refreshing in a timely manner is complicated:
+ - Unrefreshed clients and servers would be able to communicate without error
+   with both refreshed and unrefreshed clients and servers until the expiry of
+   the pre-rotation CA.
+ - Detecting clients that do not refresh would require starting a cluster with
+   automatic rotation disabled and provisioned with a near-expiry service
+   CA. This would allow a subsequent forced rotation to flush out the failure of
+   unrefreshed clients when the pre-rotation CA had expired.
+ - Without knowing that a given deployment or replicaset was automatically
+   responding to a serving cert being regenerated or a new ca bundle being
+   injected, a pod that references a serving cert secret or ca bundle configmap
+   could simply be deleted if it was started after the most recent rotation. This
```

Comment:
`deleted if it was started after the most recent rotation` - this is still about testing, not doing this in production, right? By deleting you mean `oc delete pod thispod`? If so, why delete it if it was started _after_ the most recent rotation, in that case it has the most recent artifacts for sure, no? Or is this only deletion from the set of potentially unrotated pods?

Note that in the case of just the CA bundle consumers, I would expect some of the pods to reload the certs dynamically, so there might even be pods that don't need to physically restart...

Reply:
`s/after/before/` (whoops!)

The point is that there is no way to know if a pod is using a post-rotation serving cert or ca bundle without knowing something about how it is implemented (i.e. it refreshes automatically, or knowing how to check it has refreshed).  It's a big hammer, but deleting  pods that were started before rotation (if they reference secrets or configmaps injected by the service ca operator) is the only 100% guaranteed way to ensure that all pods are using post-rotation artifacts.

Reply:
Hehe, thanks, I was not sure whether it was just a typo or something else so I asked a bunch of additional question right away in case I did not get to this today again :slightly_smiling_face:

Would it make sense to propose to try to detect replicasets/deployments with more than one pod, and delete the pods one by one in that case in order not to cause the service outage? That of course only makes sense when we find ourselves in the 6 month still-trusting the intermediate period.

Reply:
I think it could be costly to create an automated solution that ensured pods weren't using pre-rotation artifacts while also ensuring service uptime. Deployments, replicasets, daemonsets, standalone pods, etc would presumably all have to be supported and getting it wrong could cause customer pain. I'd appreciate PM input here.

Reply:
`ownerReferences` might help to some degree. This is merely pointing out the risk and proposing a possible mitigation to the customer, we are not planning our own pod-killing machine, right?

Reply:
I don't think that should be in scope, no.